### PR TITLE
Refactoring of JobsProviders and Formatters

### PR DESCRIPTION
**Pull request by patrikcerbak**
## Formatters refactoring
- migrated the tool to use the formatters from diff instead of mine
- kept the table formatters, but renamed them to `TablePrinter`

## JobsProviders refactoring
- migrated the arguments declarations of the providers into their classes (--query to JobsByQuery, --regex to JobsByRegex etc.)
- the parsing of these arguments is done in the JobsProvider classes instead of ArgumentsParsing
- extended the JobsProvider interface to have methods for:
  - getting a list of supported arguments of the provider
  - parsing a argument
  - adding a list of jobs from filesystem
  - filtering the added jobs
  - getting a list of the filtered jobs
- deleted the Providers specific parameters from the Options class, storing them instead in their classes

### CODE REVIEW

**Comment by judovana** [comment]
Both commits looks good.
Only design mini notes
  - maybe the tableformatter should go to the same package in diff where the formatters already are? 
  - maybe the old (currently used) fromatters canbe enahnced to be abel to print tbales, instead of delegating it tonew class?  I can see pros and cons for of both having them in and out :) Maybe solution which would gatehr the cons - you will have the class as implementing the TableFormatter interface as you have now. However, the old (currently used formatters) will also implement this TableFormatter, but the implementation of that method will not be here,  instead of it it will create instance of  yours current implementation of TableFormatter.
  
WDYT?
 
I was unable to determine, how the "plugins " are used at the  end - I see how you moved the query - cool - but I'm failing to find, how  it is deleagted later. Can you describe please? Can yo add links to code?

**Comment by patrikcerbak** [comment]
Also, here is a "circle of life" of the query/regex providers:
- First an instance of each is created (ArgumentsParsing:17), so the tool has the list of all supported arguments available
- Then when the arguments are parsed, and it gets an argument from any plugin (ArgumentsParsing:129), it then saves the corresponding provider to Options (ArgumentsParsing:131)
- Then the argument is given to the provider (ArgumentsParsing:141) and the provider parses it internally (e.g. JobsByQuery:60)
- Then after all arguments are parsed, the main method calls addJobs(), which adds jobs to the provider from the filesystem (VariantComparator:17, JobsByQuery:72)
- And lastly, the main method calls filterJobs() (VariantComparator:18), that filters the jobs in the provider according to the parsed arguments (JobsByQuery:76)

**Comment by patrikcerbak** [comment]
The virtual table is now appended at the bottom, so it must be used with other operation (like --compare). The rows and columns are also reversed.

**Comment by judovana** [comment]
The commit message `remade the virtual table to be ONLY available with other operations` do not align with  help message ` Can be combined with any operation, recommended to be combined with --skip-failed=false.` 

Which is true?-)

**Comment by judovana** [comment]
Btw, it do have sense to have this swithch workable also alone. But no hadrd opinion on that. Thanx!

**Comment by judovana** [comment]
So if the build found no tests, the "virtual" fields are shown anyway?

**Comment by judovana** [comment]
I see. ok!

File: src/main/java/io/jenkins/plugins/report/jtreg/main/comparator/arguments/ArgumentsDeclaration.java [hunk 1]
```diff
@@ -15,5 +15,6 @@ public final class ArgumentsDeclaration {
...
     public static final Argument onlyVolatileArg = new Argument("--only-volatile", "Specify true to show only non stable tests with the arguments list and compare (shows only tests, that are NOT failed everywhere). The default value is false.", "=<true/false>");
     public static final Argument exactTestsArg = new Argument("--exact-tests", "Specify (with regex) the exact tests to show only. The rest of tests will be ignored.", " <regex>");
-   public static final Argument formattingArg = new Argument("--formatting", "Specify the output formatting (plain, color or html). The default is plain", "=<plain/color/html>");
+    public static final Argument formattingArg = new Argument("--formatting", "Specify the output formatting (plain, color or html). The default is plain.", "=<plain/color/html>");
+    public static final Argument useDefaultBuildArg = new Argument("--use-default-build", "If set to true and no matching build with given NVR was found, the tool will use the latest (default) build instead. Default value is false.", "=<true/false>");
```
Comment (judovana) [inline src/main/java/io/jenkins/plugins/report/jtreg/main/comparator/arguments/ArgumentsDeclaration.java:19]: There is an uhappy manual synchronization - the argument have --name and description here, however the default value is set elsewhere.  And in addition, the default value is hardwritten in help sting. This is wide spread over allargs. Most likley not neessary to fix here, but somehting to keep in mind for any future works.

Comment (judovana) [reply src/main/java/io/jenkins/plugins/report/jtreg/main/comparator/arguments/ArgumentsDeclaration.java:19]: You may also see, that with switches  with enumerations, this got complicated in addition  by hardwritten enumeration Eg <plain/color/html> (but not much differences for <true/false> actually)

File: src/main/java/io/jenkins/plugins/report/jtreg/main/diff/formatters/ColorFormatter.java [hunk 2]
```diff
@@ -122,4 +123,51 @@ public void startTitle4() {
         startColor(SupportedColors.Blue);
     }
 
+    @Override
+    public void printTable(String[][] table, int rowSize, int columnSize) {
+        // first print the first row definitions
```
Comment (judovana) [inline src/main/java/io/jenkins/plugins/report/jtreg/main/diff/formatters/ColorFormatter.java:128]: What is the reason the columnsize have to be set, and not deducted by method?

**Pull request merged**